employee_api/handlers/employee.go

```diff
@@ -0,0 +1,246 @@
+package employee_handler
+
+import (
+	"HR-system/employee_api/models"
+	responses "HR-system/employee_api/utils"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+	"text/template"
+
+	"github.com/gorilla/mux"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+type Route struct {
+	Name        string
+	Method      string
+	Pattern     string
+	HandlerFunc http.HandlerFunc
+}
+
+type Server struct {
+	DB     *gorm.DB
+	Router *mux.Router
+	Routes []Route
+}
+
+func (s *Server) Run(
+	host, port, user, dbname, password, sslmode string,
+) {
+	s.initializeDB(host, port, user, dbname, password, sslmode)
+	s.startServe(":8080")
+}
+
+func (s *Server) collectRoutes() []Route {
+	return []Route{
+		{
+			"Index",
+			"GET",
+			"/",
+			s.index,
+		},
+		{
+			"Create employee",
+			"POST",
+			"/api/v1/employee/",
+			s.create,
+		},
+		{
+			"Get employee by id",
+			"GET",
+			"/api/v1/employee/{id}",
+			s.get,
+		},
+		{
+			"Update employee",
+			"PUT",
+			"/api/v1/employee/{id}",
+			s.update,
+		},
+		{
+			"Delete employee",
+			"DELETE",
+			"/api/v1/employee/{id}",
+			s.delete,
+		},
+	}
+
+}
+
+func (server *Server) startServe(addr string) {
+	log.Println("Listening to port 8080")
+	log.Fatal(http.ListenAndServe(addr, server.Router))
+}
+
+func (s *Server) initializeDB(host, port, user, dbname, password, sslmode string) {
+	databaseUrl := fmt.Sprintf("host=%s port=%s user=%s dbname=%s password=%s sslmode=%s", host, port, user, dbname, password, sslmode)
+	db, err := gorm.Open(postgres.Open(databaseUrl), &gorm.Config{})
+
+	if err != nil {
+		log.Printf("Cannot connect to %s database", dbname)
+		log.Fatal("This is the error:", err)
+		return
+	}
+
+	s.DB = db
+
+	s.DB.AutoMigrate(&models.Employee{})
+
+	s.Router = mux.NewRouter()
+	s.Routes = s.collectRoutes()
+	for _, route := range s.Routes {
+		s.Router.HandleFunc(route.Pattern, route.HandlerFunc).Methods(route.Method)
+	}
+}
+
+func (s *Server) create(w http.ResponseWriter, r *http.Request) {
+	// read data from request body
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		responses.Error(w, http.StatusBadRequest, []string{"Invalid request"})
+		return
+	}
+
+	// convert data to Employee struct
+	employee := models.Employee{}
+	err = json.Unmarshal(body, &employee)
```

Comment:
usually, I don't prefer to parse directly the data from the request body into the database model, it could include some unexpected field (eg: id,...). So I suggest using another object as a DTO struct then we map field by field (and maybe validate data before mapping).

Reply:
did you mean binding json like this
type Employee struct {
	ID        int    `json:"id" binding:"required,number" gorm:"primary_key"`
	FirstName string `json:"first_name" binding:"required"`
	LastName  string `json:"last_name" binding:"required"`
	Email     string `json:"email" binding:"required,email"`
	Phone     string `json:"phone" binding:"required,number"`
	Birthday  string `json:"birthday" binding:"required,datetime=2006-01-02"`
	StartDay  string `json:"start_day" binding:"required,datetime=2006-01-02"`
	Position  string `json:"position" binding:"required,oneof=developer tester manager"`
}
`
`